### PR TITLE
Customer success Platform Backend.

### PR DESCRIPTION
**Pull request by Rahul-Maity**

### CODE REVIEW

File: Promact.CustomerSuccess.Platform/Controllers/AccountController.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+//                // Note that the resulting absolute Uri must be added to the
+//                // **Allowed Logout URLs** settings for the app.
+//                .WithRedirectUri(Url.Action("Index", "Home"))
+//                .Build();
+
+//            await HttpContext.SignOutAsync(Auth0Constants.AuthenticationScheme, authenticationProperties);
+//            await HttpContext.SignOutAsync(CookieAuthenticationDefaults.AuthenticationScheme);
+//        }
+//    }
+//}
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Controllers/AccountController.cs:37]: Commented code should not be in PR,

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs [hunk 2]
```diff
@@ -0,0 +1,958 @@
...
+            email.From.Add(MailboxAddress.Parse("mrahulmaity623@gmail.com"));
+            email.To.Add(MailboxAddress.Parse(userEmail));
+            email.Subject = subject;
+            email.Body = new TextPart(MimeKit.Text.TextFormat.Html)
+            {
+                Text = body
+            };
+
+            using var smtp = new SmtpClient();
+            smtp.Connect("smtp.gmail.com", 587, SecureSocketOptions.StartTls);
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Controllers/EmailController.cs:874]: Don't put sensitive details in code

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs [hunk 3]
```diff
@@ -0,0 +1,958 @@
...
+    public class EmailRequestDocumentVersion
+    {
+        public List<Stakeholder> Stakeholders { get; set; }
+        //public AuditHistory AuditData { get; set; }
+        public DocumentVersion versionData { get; set; }
+    }
+
+    public class EmailRequestProjectScope
+    {
+        public List<Stakeholder> Stakeholders { get; set; }
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Controllers/EmailController.cs:944]: same as above

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs [hunk 4]
```diff
@@ -0,0 +1,958 @@
...
+        //public ApprovedTeam ApprovedTeam { get; set; }
+    }
+
+    public class EmailRequestApprovedTeam
+    {
+        public List<Stakeholder> Stakeholders { get; set; }
+        //public AuditHistory AuditData { get; set; }
+        public ApprovedTeam ApprovedData { get; set; }
+    }
+    public class EmailRequestBudget
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Controllers/EmailController.cs:893]: Don't put classes randomly

File: Promact.CustomerSuccess.Platform/Entities/Project.cs [hunk 5]
```diff
@@ -12,7 +13,10 @@ public class Project : AuditedEntity<Guid>
         public virtual ICollection<RiskProfile>? RiskProfiles { get; set; }
         public virtual ICollection<PhaseMilestone>? PhaseMilestones { get; set; }
         public virtual ICollection<ProjectResources>? Resources { get; set; }
-        public virtual ICollection<ClientFeedback>? ClientFeedbacks { get; set; }
+        //public virtual ICollection<ProjectResources>? ProjectResources { get; set; }
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Entities/Project.cs:16]: Commented code should not be in PR,

**Review by rahul-COD3 — CHANGES_REQUESTED** [review]

File: Promact.CustomerSuccess.Platform/Services/DocumentVersionService.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+using Volo.Abp.Application.Dtos;
+using Volo.Abp.Application.Services;
+using Volo.Abp.Domain.Repositories;
+
+namespace Promact.CustomerSuccess.Platform.Services
+{
+    public class DocumentVersionService:
+
+        CrudAppService<DocumentVersion,
+            DocumentVersionDto,
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Services/DocumentVersionService.cs:13]: Fix the formatting

File: Promact.CustomerSuccess.Platform/Entities/Sprint.cs [hunk 7]
```diff
@@ -1,22 +1,22 @@
 ﻿using System.ComponentModel.DataAnnotations.Schema;
+using Volo.Abp.Domain.Entities;
 using Volo.Abp.Domain.Entities.Auditing;
 
 namespace Promact.CustomerSuccess.Platform.Entities
 {
-    public class Sprint : AuditedAggregateRootWithUser<Guid, ApplicationUser>
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Entities/Sprint.cs:6]: Why you removed  AuditedAggregateRootWithUser<Guid, ApplicationUser>

File: Promact.CustomerSuccess.Platform/Services/Dtos/CreateProjectResourcesDto.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Promact.CustomerSuccess.Platform.Services.Dtos
+{
+    public class CreateProjectResourcesDto
+    {
+        public Guid ProjectId { get; set; }
+
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Services/Dtos/CreateProjectResourcesDto.cs:6]: Fix formatting

File: Promact.CustomerSuccess.Platform/Controllers/AccountController.cs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    public class AccountController : Controller
+    {
+        public async Task Login(string returnUrl = "/")
+        {
+            var authenticationProperties = new LoginAuthenticationPropertiesBuilder()
+
+                .WithRedirectUri(returnUrl)
+                .Build();
+
+            await HttpContext.ChallengeAsync(Auth0Constants.AuthenticationScheme, authenticationProperties);
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Controllers/AccountController.cs:60]: Use dependency injection for accessing HttpContext

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs [hunk 10]
```diff
@@ -0,0 +1,958 @@
...
+        public IActionResult SendEmail([FromBody] EmailRequest request)
+        {
+            if (!ModelState.IsValid) { return BadRequest(ModelState); }
+
+
+            //var emailBody = getEmailBody(auditData);
+            foreach (var stakeholder in request.Stakeholders)
+            {
+                var emailBody = getEmailBody(stakeholder, request.AuditData);
+                SendEmail(stakeholder.Contact, "Audit History Updated", emailBody);
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Controllers/EmailController.cs:31]: Lack of async await

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs [hunk 11]
```diff
@@ -0,0 +1,958 @@
...
+        [HttpPost("sendEmailApprovedTeam")]
+        public IActionResult sendEmailApprovedTeam([FromBody] EmailRequestApprovedTeam request)
+        {
+            if (!ModelState.IsValid) { return BadRequest(ModelState); }
+
+
+            foreach (var stakeholder in request.Stakeholders)
+            {
+                var emailBody = getEmailBodyApprovedTeam(stakeholder, request.ApprovedData);
+                SendEmail(stakeholder.Contact, "ApprovedTeam History Updated", emailBody);
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Controllers/EmailController.cs:54]: Lack of async await

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs [hunk 12]
```diff
@@ -0,0 +1,958 @@
...
+        [HttpPost("sendEmailProjectScope")]
+        public IActionResult sendEmailProjectScope([FromBody] EmailRequestProjectScope request)
+        {
+            if (!ModelState.IsValid) { return BadRequest(ModelState); }
+
+
+            foreach (var stakeholder in request.Stakeholders)
+            {
+                var emailBody = getEmailBodyProjectScope(stakeholder, request.scopeData);
+                SendEmail(stakeholder.Contact, "Project Scope Updates", emailBody);
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Controllers/EmailController.cs:192]: same as above

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs [hunk 13]
```diff
@@ -0,0 +1,958 @@
...
+        [HttpPost("sendEmailPhaseMileStone")]
+        public IActionResult sendEmailPhaseMileStone([FromBody] EmailRequestPhaseMileStone request)
+        {
+            if (!ModelState.IsValid) { return BadRequest(ModelState); }
+
+
+            foreach (var stakeholder in request.Stakeholders)
+            {
+                var emailBody = getEmailBodyPhaseMilestone(stakeholder, request.phaseMileStoneData);
+                SendEmail(stakeholder.Contact, "Project Phase Milestone Updates", emailBody);
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Controllers/EmailController.cs:207]: same as above

File: Promact.CustomerSuccess.Platform/Entities/PhaseMilestone.cs [hunk 14]
```diff
@@ -14,7 +14,7 @@ public class PhaseMilestone : AuditedAggregateRootWithUser<Guid, ApplicationUser
         public required string Comments { get; set; }
         public MilestoneOrPhaseStatus Status { get; set; }
         public virtual Project? Project { get; set; }
-        public virtual ICollection<Sprint>? Sprints { get; set; }
+        //public virtual ICollection<Sprint>? Sprints { get; set; }
```
Comment (rahul-COD3) [inline Promact.CustomerSuccess.Platform/Entities/PhaseMilestone.cs:17]: Why did you remove the public virtual ICollection<Sprint> Sprints { get; set; }?

**Comment by Rahul-Maity** [comment]
fixed formatting and integrated email in service.

**Review by Rahul-Maity — COMMENTED** [review]
Email service

File: Promact.CustomerSuccess.Platform/Controllers/AccountController.cs (hunk 1 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/AccountController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/AccountController.cs [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    //            // Redirect the user back to the Angular frontend
+    //            return Redirect("http://localhost:4200/");
+    //        }
+    //        catch (Exception ex)
+    //        {
+    //            return StatusCode(500, $"Internal server error: {ex.Message}");
+    //        }
+    //    }
+    //}
+}
```
[the remark this answers is not in the document]
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/AccountController.cs:46]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs [hunk 16]
```diff
@@ -0,0 +1,958 @@
...
+namespace Promact.CustomerSuccess.Platform.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class EmailController : ControllerBase
+    {
+        [HttpPost("sendEmail")]
+        public IActionResult SendEmail([FromBody] EmailRequest request)
+        {
+            if (!ModelState.IsValid) { return BadRequest(ModelState); }
```
[the remark this answers is not in the document]
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/EmailController.cs]: removed controller

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs [hunk 17]
```diff
@@ -0,0 +1,958 @@
...
+    {
+        public List<Stakeholder> Stakeholders { get; set; }
+        //public AuditHistory AuditData { get; set; }
+        public ProjectUpdate projectUpdateData { get; set; }
+    }
+
+    public class EmailRequestProjectDocument
+    {
+        public List<Stakeholder> Stakeholders { get; set; }
+        //public AuditHistory AuditData { get; set; }
```
[the remark this answers is not in the document]
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/EmailController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/HomeController.cs [hunk 18]
```diff
@@ -8,5 +8,7 @@ public class HomeController : AbpController
     public ActionResult Index()
     {
         return Redirect("~/swagger");
+        //return Redirect("http://localhost:4200/");
```
[the remark this answers is not in the document]
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/HomeController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Entities/Sprint.cs [hunk 19]
```diff
@@ -1,22 +1,22 @@
...
-        public int SprintNumber { get; set; }
-        public virtual PhaseMilestone? PhaseMilestone { get; set; }
-        public override object?[] GetKeys()
-        {
-            throw new NotImplementedException();
-        }
+        public string Comments { get; set; }
+
+    
+
```
[the remark this answers is not in the document]
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Entities/Sprint.cs:20]: Fixed

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs (hunk 2 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/EmailController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs (hunk 3 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/EmailController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs (hunk 4 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/EmailController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Entities/Project.cs (hunk 5 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Entities/Project.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Services/DocumentVersionService.cs (hunk 6 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Services/DocumentVersionService.cs:16]: Fixed formatting

File: Promact.CustomerSuccess.Platform/Entities/Sprint.cs (hunk 7 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Entities/Sprint.cs:6]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/AccountController.cs (hunk 9 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/AccountController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs (hunk 10 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/EmailController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs (hunk 11 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/EmailController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs (hunk 12 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/EmailController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Controllers/EmailController.cs (hunk 13 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Controllers/EmailController.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Entities/PhaseMilestone.cs (hunk 14 above)
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Entities/PhaseMilestone.cs]: Resolved

File: Promact.CustomerSuccess.Platform/Services/EscalationMatrixService.cs [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+using Volo.Abp.Domain.Repositories;
+
+
+namespace Promact.CustomerSuccess.Platform.Services
+{
+    public class EscalationMatrixService :
+
+        CrudAppService<EscalationMatrix,
+            EscalationMatrixDto,
+
```
[the remark this answers is not in the document]
Comment (Rahul-Maity) [reply Promact.CustomerSuccess.Platform/Services/EscalationMatrixService.cs]: Resolved

**Pull request closed without merging**